### ISSUE TITLE
New rule: Fix invalid cast for (ancient) pre-generic collection types

### ISSUE THREAD

**Issue by sos-dll** [dotnet/roslyn-analyzers#4067]
## Overview

Fix `InvalidCastException` at compile-time rather than during run-time.
This is analogous to "having C# without type-checker".
It is to do with (old) pre-generic collections such as: `MatchCollection`, `ProcessModuleCollection`, etc (there might be even more of those types, most likely).

## Suggestion

All of those (old) collection types can only be bound to their actual *backing type* (or `object`, also using `var` makes it an `object` as well).
In other words, for--
- `System.Text.RegularExpressions.MatchCollection` use `System.Text.RegularExpressions.Match`
- `System.Diagnostics.ProcessModuleCollection` use `System.Diagnostics.ProcessModule`
- etc.

So, if the type is neither kind of *backing type* nor `System.Object` nor `dynamic`, a compiler should report an error; otherwise, an exception is going to raise at runtime.
Optionally: Perhaps also consider warning when using `object`/`var`, since then you can't access data/properties without casting it to the correct *backing type* or by using `dynamic` (but it may not be desirable most of the time that is, a code smell as seen from my point of view).

## Additional context

```cs
// The following two are fine, except in my opinion I would prefer to have warning generated.
foreach (var/*object*/ module in Process.GetCurrentProcess().Modules) { Console.WriteLine(module); }
foreach (dynamic match in new Regex("foo").Matches("...foo's bar...")) { Console.WriteLine(match.Index); }

// Accidentally ended up using System.Reflection.Module - realized at runtime (too late).
foreach (Module module in Process.GetCurrentProcess().Modules) { Console.WriteLine(module); }

// Very easy to slip and make this silly mistake without actual type-hinting.
foreach (string match in new Regex("foo").Matches("...foo's bar...")) { Console.WriteLine(match); }
```

Do I need to say more? Did I miss anything? 📑
Side exploration: Fortunately, for ReSharper and probably Rider users as well, if you use `.foreach` (postfix) template to generate the foreach loop, it will automatically figure out and use the correct *backing type* initially (afterwards, not even smart intellisense can figure out the correct *backing type* if you have changed the data-type, so don't try to change that part after it has been generated out for you).

P.S. I am a bit confused, idk if I should have posted this under runtime repo instead, anyways please move the issue where it is fit (if not here).

----

**Category**: Reliability (correct me if i am wrong)

**Comment by svick** [comment]
For the record, [`MatchCollection`](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.matchcollection?view=netcore-2.0) is kind of generic since .Net Core 2.0: it implements `IList<Match>`, though simple `foreach` over it still uses `object` (I think that's because of backwards compatibility).

You can make it work better in `foreach` if you add `.AsEnumerable()`:

```c#
// CS0030 Cannot convert type 'System.Text.RegularExpressions.Match' to 'string'
foreach (string match in new Regex("foo").Matches("...foo's bar...").AsEnumerable()) { Console.WriteLine(match); }
```

Maybe the warning (if it's actually worth implementing) could figure out what the "backing type" is based on the implemented interfaces?

**Comment by Youssef1313** [comment]
Analyzer for `foreach` implicit casts was added in dotnet/roslyn (IDE0220). I think this can be closed now?

**Comment by mavasani** [comment]
Yes sounds reasonable
